### PR TITLE
Use `PyccelAstNode.pyccel_staging` in `SemanticParser`

### PR DESCRIPTION
**Pull request by EmilyBourne**
`PyccelAstNode`s have a `pyccel_staging` property. This is not used which makes it hard to detect when it is set incorrectly. Separately we often want to create a partially syntactic object (e.g. when unravelling an assignment where the `rhs` has been visited and the `lhs` is an inhomogeneous tuple). This PR uses the `pyccel_staging` property to facilitate the implementation of the latter situation. This is done by exiting `SemanticParser._visit` immediately if the object passed to the function is already a semantic object. As the property is now used it is obvious when it has not been correctly set. These cases are therefore fixed

### CODE REVIEW

**Review by Farouk-Echaref — APPROVED** [review]
looks good, i just want to know in which case exactly this pyccel_staging managed to break the code.

**Comment by EmilyBourne** [comment]
> looks good, i just want to know in which case exactly this pyccel_staging managed to break the code.

@Farouk-Echaref After this PR, if `pyccel_staging` is incorrectly set to semantic the node is not visited. This means that it doesn't contain any type information. Errors quickly follow when you try to do anything with the node

File: pyccel/parser/semantic.py [hunk 1]
```diff
@@ -1933,6 +1948,8 @@ def _visit(self, expr):
         pyccel.ast.basic.PyccelAstNode
             AST object which is the semantic equivalent of expr.
         """
+        if getattr(expr, 'pyccel_staging', 'syntactic') == 'semantic':
+            return expr
```
Comment (yguclu) [inline pyccel/parser/semantic.py:1952]: Why using `getattr` with a default value if `expr` is of type `PyccelAstNode`, which has the `pyccel_staging` property?

**Review by yguclu — CHANGES_REQUESTED** [review]

File: pyccel/parser/semantic.py (hunk 1 above)
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1952]: Because sometimes `expr` is a `PyccelSymbol` which isn't a `PyccelAstNode`

Comment (yguclu) [reply pyccel/parser/semantic.py:1952]: The docstring says otherwise...

Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1952]: Fixed 18c473a6f54633e9e3d52ec60e5a576fee28ca2d

**Review by yguclu — APPROVED** [review]

**Pull request merged**